lib/Echidna/RPC.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Echidna.RPC where
+
+import Prelude hiding (Word)
+
+import Control.Exception (Exception)
+import Control.Lens
+import Control.Monad (foldM)
+import Control.Monad.Catch (MonadThrow, throwM)
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Reader.Class (MonadReader(..))
+import Control.Monad.State.Strict (MonadState, execStateT, runStateT, get, put, runState)
+import Data.Aeson (FromJSON(..), (.:), withObject, eitherDecodeFileStrict)
+import Data.Binary.Get (runGetOrFail)
+import Data.ByteString.Char8 (ByteString, empty)
+import Data.Has (Has(..))
+import Data.List (partition)
+import Data.Map (fromList)
+import Data.Text.Encoding (encodeUtf8)
+import EVM
+import EVM.ABI (AbiType(..), getAbi)
+import EVM.Concrete (w256)
+import EVM.Exec (exec, vmForEthrunCreation)
+import EVM.Types (Addr, W256)
+
+import qualified Control.Monad.Fail as M (MonadFail(..))
+import qualified Control.Monad.State.Strict as S (state)
+import qualified Data.ByteString.Base16 as BS16 (decode)
+import qualified Data.Text as T (Text, drop)
+import qualified Data.Vector as V (fromList)
+
+import Echidna.Exec
+import Echidna.Transaction
+
+-- | During initialization we can either call a function or create an account or contract
+data Etheno = AccountCreated Addr
+            | ContractCreated Addr Addr Integer Integer ByteString W256
+            | FunctionCall Addr Addr Integer Integer ByteString W256
+  deriving (Eq, Show)
+
+instance FromJSON Etheno where
+  -- parseJSON = genericParseJSON $ defaultOptions {fieldLabelModifier = tail, omitNothingFields = True}
+  parseJSON = withObject "Etheno" $ \v -> do
+    (ev :: String) <- v .: "event"
+    case ev of
+         "AccountCreated"  -> AccountCreated  <$> v .: "address"
+         "ContractCreated" -> ContractCreated <$> v .: "from"
+                                              <*> v .: "contract_address"
+                                              <*> (read <$> (v .: "gas_used"))
+                                              <*> (read <$> (v .: "gas_price"))
+                                              <*> (decode <$> (v .: "data"))
+                                              <*> v .: "value"
+         "FunctionCall"    -> FunctionCall    <$> v .: "from"
+                                              <*> v .: "to"
+                                              <*> (read <$> (v .: "gas_used"))
+                                              <*> (read <$> (v .: "gas_price"))
+                                              <*> (decode <$> (v .: "data"))
+                                              <*> v .: "value"
+         _ -> M.fail "event should be one of \"AccountCreated\", \"ContractCreated\", or \"FunctionCall\""
+    where decode :: T.Text -> ByteString
+          decode = decodeHex . encodeUtf8 . T.drop 2
+          decodeHex = fst . BS16.decode
```

Comment:
`hexstring`'s interface can bottom. `BS16.decode` decodes as much as it can in the `fst` element and puts the rest in `snd`. `drop`ping from empty strings just yields the empty string so this is actually not partial.

Reply:
hmmm, does this still fail silently then? I think that for any user-provided field, we should fail loudly if we fail to parse the full input given

Reply:
no longer fails silently